### PR TITLE
Added automatic latency measurement

### PR DESCRIPTION
**Pull request by alejandroliu**
Hello,

This PR adds automatic ping checking to RakLib.  It does so by sending a ping after the client intiated ping/pong sequence is handled.  It then measures the returning pong and saves that.

Cheers

### CODE REVIEW

File: server/Session.php [hunk 1]
```diff
@@ -387,7 +392,21 @@ private function handleEncapsulatedPacketRoute(EncapsulatedPacket $packet){
...
+				$pk->encode();
+				$sendPacket = new EncapsulatedPacket();
+				$sendPacket->reliability = 0;
+				$sendPacket->buffer = $pk->buffer;
+				$this->addToQueue($sendPacket);
+			}elseif($id === PONG_DataPacket::$ID){
+				$dataPacket = new PONG_DataPacket;
+				$dataPacket->buffer = $packet->buffer;
+				$dataPacket->decode();
+				$this->lastPing = (intval(\microtime(\true) * 10000) - $dataPacket->pingID) / 10.0;
```
Comment (PEMapModder) [inline server/Session.php]: Actually you don't really need the `\\` here

**Comment by alejandroliu** [comment]
Just trying to keep the same coding style as elsewhere...

**Comment by shoghicp** [comment]
@alejandroliu Please check the original source code, that kind of formatting is added by the second step of code optimization after the preprocessor inlining.

**Comment by alejandroliu** [comment]
Is there another source code?  I though everything was in the github repository.

**Comment by alejandroliu** [comment]
Removed the \

**Comment by alejandroliu** [comment]
Anything else needed to get this pull request accepted?

**Comment by alejandroliu** [comment]
OK, I added more code so that the latency is measured and reported to the main thread by a Ping Packet.  I have some code (against PocketMine 1.5dev) that can be used to track the Ping values which I can add as a separate PullRequest.

In the meantime, it is possible to track the latency measurements through a Plugin as this should trigger a DataPacketReceiveEvent with an UnknownPacket but you can check the pid and get the millisecond latency from the buffer.

**Comment by shoghicp** [comment]
@alejandroliu I'll check this later, and maybe add the proper details to the RakLib interfaces and AdvancedSourceInterface on pocketmine (so other people can implement it as well!)

**Comment by alejandroliu** [comment]
I created a simple plugin [GetPing](https://github.com/alejandroliu/bad-plugins/releases/tag/GetPing-1.0.0) to demonstrate the use of this PullRequest.  No changes to the PocketMine core were needed.

Obviously this is a quick and dirty implementation that only reports last measurements.  It is possible to record more measurements and averages.

**Pull request closed without merging**